hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/SigTransactionMetadata.java

```diff
@@ -84,8 +78,94 @@ public ResponseCodeEnum status() {
         this.status = status;
     }
+
+    /**
+     * Checks if the accountId is same as payer or the status of the metadata is already failed. If
+     * either of the above is true, doesn't look up the keys for given account. Else, looks up the
+     * keys for account. If the lookup fails, sets the default failureReason given in the result.
+     *
+     * @param id given accountId
+     */
+    public void addNonPayerKey(final AccountID id) {
+        addNonPayerKey(id, null);
+    }
+
+    /**
+     * Checks if the accountId is same as payer or the status of the metadata is already failed. If
+     * either of the above is true, doesn't look up the keys for given account. Else, looks up the
+     * keys for account. If the lookup fails, sets the given failure reason on the metadata. If the
+     * failureReason is null, sets the default failureReason given in the result.
+     *
+     * @param id given accountId
+     * @param failureStatus given failure status
+     */
+    public void addNonPayerKey(final AccountID id, final ResponseCodeEnum failureStatus) {
+        if (isNotNeeded(id)) {
+            return;
+        }
+        final var result = store.getKey(id);
+        failOrAddToKeys(result, failureStatus);
+    }
+
+    /**
+     * Checks if the accountId is same as payer or the status of the metadata is already failed. If
+     * either of the above is true, doesn't look up the keys for given account. Else, looks up the
+     * keys for account if receiverSigRequired is true on the account. If the lookup fails, sets the
+     * given failure reason on the metadata. If the failureReason is null, sets the default
+     * failureReason given in the result.
+     *
+     * @param id given accountId
+     * @param failureStatus given failure status
+     */
+    public void addNonPayerKeyIfReceiverSigRequired(
+            final AccountID id, final ResponseCodeEnum failureStatus) {
+        if (isNotNeeded(id)) {
+            return;
+        }
+        final var result = store.getKeyIfReceiverSigRequired(id);
+        failOrAddToKeys(result, failureStatus);
+    }
+
+    /* ---------- Helper methods ---------- */
+
+    /**
+     * Look up the keys for payer account and add payer key to the required keys list. If the lookup
+     * fails adds failure status {@code INVALID_PAYER_ACCOUNT_ID} to the metadata.
+     */
+    private void addPayerKey() {
+        final var result = store.getKey(payer);
+        failOrAddToKeys(result, INVALID_PAYER_ACCOUNT_ID);
+    }
+
+    /**
+     * Checks if the account given is same as payer or if the metadata is already failed. In either
+     * case, no need to lookup tha account's key.
+     *
+     * @param id given account
+     * @return true if the lookup is not needed, false otherwise
+     */
+    private boolean isNotNeeded(final AccountID id) {
+        return id.equals(payer) || id.equals(AccountID.getDefaultInstance()) || status != OK;
```

Comment:
This `id.equals(AccountID.getDefaultInstance())` exclusion is specific to `AccountID`'s set in the `CryptoAllowance#owner`, `NftAllowance#owner`, and `TokenAllowance#owner` fields, right?

So on one hand, it seems like we should explicitly limit this exemption to those three contexts.

But on the other hand, if someone leaves a required `AccountID` unset, how could the transaction succeed at consensus? 😁 We're only going to touch the payer account for node + network fees anyways, why bother checking any other signatures?

The same argument would then apply to the `0.0.0` sentinel id, and we could make this:
```
return id.equals(payer)
    || id.equals(AccountID.getDefaultInstance())
    || designatesAccountRemoval(id)
    || status != OK;

...

    private boolean designatesAccountRemoval(AccountID id) {
        return id.getShardNum() == 0
                && id.getRealmNum() == 0
                && id.getAccountNum() == 0
                && id.getAlias().isEmpty();
    }
```

Reply:
> This id.equals(AccountID.getDefaultInstance()) exclusion is specific to AccountID's set in the CryptoAllowance#owner, NftAllowance#owner, and TokenAllowance#owner fields, right?

yeah , I am not sure, why this condition is added only for allowances ? 🤔 

> The same argument would then apply to the 0.0.0 sentinel id, and we could make this:

Agreed